### PR TITLE
Debug/add transfer func to protocols

### PR DESCRIPTION
**Pull request by farnoodmah**
Add LinuxIpc-Issues document.
Add timelimit feature to protocols.
Add test cases.

### CODE REVIEW

File: cpp_exercise/src/lib/ipcmsgqlib.cc [hunk 1]
```diff
@@ -11,23 +11,36 @@
...
 
+  
         //removing and unlinking from old shared memories
 
-        remove(_msg_queue_name.c_str());
-        ret = mq_unlink(_msg_queue_name.c_str());
+
+        remove(msg_queue_name.c_str());
+        ret = mq_unlink(msg_queue_name.c_str());
+
```
Comment (KinhTruongUnikie) [inline cpp_exercise/src/lib/ipcmsgqlib.cc:21]: I don't think you need remove() here, mq_unlink() will remove the message queue name

File: cpp_exercise/src/lib/ipcmsgqlib.cc [hunk 2]
```diff
@@ -61,9 +73,27 @@ MsgQueueSender::MsgQueueSender(const std::string & filename): _file_name(filenam
...
   
 
-            mq_send(_msg_queue, reinterpret_cast<char *>(buffer.data()),buffer.size(),_priority);
+           
+            ret = mq_timedsend(_msg_queue,(char*)buffer.data(),buffer.size(),_priority, &_ts);
+
+            if(ret<0){
+                if (errno == ETIMEDOUT){
+                std::cout<<"          "<<strerror(errno)<<std::endl;
+                throw IPCException("          PCSender ERROR: Cannot Connect to the Receiver");
```
Comment (KinhTruongUnikie) [inline cpp_exercise/src/lib/ipcmsgqlib.cc:89]: You can put `strerror(errno)` inside throw(), errno will be printed together with the error message that way

**Review by baovu-unikie — APPROVED** [review]
Good work! 

There are a lot of multiple blank lines in the code.

File: cpp_exercise/src/lib/ipcmsgqlib.cc (hunk 1 above)
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcmsgqlib.cc:21]: Fixed

File: cpp_exercise/src/lib/ipcmsgqlib.cc (hunk 2 above)
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcmsgqlib.cc:89]: Fixed

File: cpp_exercise/src/lib/ipcmsgqlib.cc [hunk 3]
```diff
@@ -74,6 +104,24 @@ MsgQueueSender::MsgQueueSender(const std::string & filename): _file_name(filenam
...
+
+            }
+                std::cout<<"          Checking the IPCReceiver received the file completely... Try:"<<counter<<"/5"<<std::endl;
+                 mq_getattr(_msg_queue, &_attrs);
+	 	 		_check_empty = _attrs.mq_curmsgs;
+
+                if(_check_empty == 0){
+                    break;
+                }
+                sleep(5);
```
[the remark this answers is not in the document]
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcmsgqlib.cc:122]: Fixed

File: cpp_exercise/src/lib/ipcmsgqlib.cc [hunk 4]
```diff
@@ -11,23 +11,36 @@
...
-        std::cout<<"**************************************************************"<<std::endl;
 
+  
         //removing and unlinking from old shared memories
 
-        remove(_msg_queue_name.c_str());
-        ret = mq_unlink(_msg_queue_name.c_str());
+
+        remove(msg_queue_name.c_str());
+        ret = mq_unlink(msg_queue_name.c_str());
```
[the remark this answers is not in the document]
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcmsgqlib.cc:20]: Fixed

File: cpp_exercise/src/lib/ipcmsgqlib.cc [hunk 5]
```diff
@@ -39,8 +52,7 @@ MsgQueueSender::MsgQueueSender(const std::string & filename): _file_name(filenam
 
     
 
-        //opening the msg queue
-        _msg_queue = mq_open(_msg_queue_name.c_str(), O_WRONLY | O_CREAT , S_IRWXU | S_IRWXG, &_attrs);
+
 
 
         if(_msg_queue<0){
```
[the remark this answers is not in the document]
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcmsgqlib.cc:58]: Fixed

File: cpp_exercise/src/lib/ipcmsgqlib.cc [hunk 6]
```diff
@@ -74,6 +104,24 @@ MsgQueueSender::MsgQueueSender(const std::string & filename): _file_name(filenam
...
+
+                if(_check_empty == 0){
+                    break;
+                }
+                sleep(5);
+        }
+
       
 
        std::cout<<"          Successfully written to the MSGQueue "<<std::endl;
```
[the remark this answers is not in the document]
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcmsgqlib.cc:127]: Fixed

File: cpp_exercise/src/lib/ipcmsgqlib.cc [hunk 7]
```diff
@@ -109,19 +160,31 @@ MsgQueueReceiver::MsgQueueReceiver(const std::string & filename): _file_name(fil
...
+            std::cout<<"          Waiting for the Sender...Try:"<<counter<<"/5"<<std::endl;     
+            _msg_queue = mq_open(msg_queue_name.c_str(), O_RDONLY , &_attrs); 
 
             if(_msg_queue>=0){
 
                  mq_getattr(_msg_queue, &_attrs);
 	 	 		_check_empty = _attrs.mq_curmsgs;
             }
+            
             sleep(5);
```
[the remark this answers is not in the document]
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcmsgqlib.cc:179]: Fixed

File: cpp_exercise/src/lib/ipcpipelib.cc [hunk 8]
```diff
@@ -106,71 +129,105 @@ PipeSender::~PipeSender(){
...
+      do{
+        if (0>(_fifo=open(_myfifo,O_RDONLY)))
+            if(errno==EINTR){
+                if(abort_eh){
+                  throw IPCException("IPCSender ERROR: Cannot to the IPCSender.");
+                }  
+                else{
+                  continue; 
+                } 
+            }
```
[the remark this answers is not in the document]
Comment (farnoodmah) [reply cpp_exercise/src/lib/ipcpipelib.cc:179]: Fixed

**Review by KinhTruongUnikie — APPROVED** [review]

**Pull request merged**